### PR TITLE
Add update feature from api-gateway

### PR DESCRIPTION
**Pull request by yosef32**
Signed-off-by: yosef32 <yosef32@gmail.com>
close issue #200

### CODE REVIEW

File: src/upload/upload.service.ts [hunk 1]
```diff
@@ -51,6 +52,42 @@ export class UploadService {
...
+    ) : Promise<IUpload> {
+
+    const file: IFile = await FilesRepository.getFileInFolderByName(parent, name, ownerID);
+    if (!file) {
+      throw new FileNotFoundError();
+    }
+
+    let sizeCalculated: number = 0;
+    if (file.size > size) {
+      sizeCalculated = file.size - size;
```
Comment (Shahar-Y) [inline src/upload/upload.service.ts:80]: Hafuh, Guta, Hafuh

File: src/file/file.service.spec.ts [hunk 2]
```diff
@@ -154,6 +154,50 @@ describe('File Logic', () => {
     });
   });
 
+  describe('#createUpdate', () => {
+    it('should return a new upload', async () => {
+      const key: string = KEY;
+      const key2: string = KEY;
+      const key3: string = KEY;
```
Comment (Shahar-Y) [inline src/file/file.service.spec.ts:168]: Why three identical keys? and one of them is used as parentID? 
just use the global `KEY`

File: src/file/file.service.spec.ts [hunk 3]
```diff
@@ -154,6 +154,50 @@ describe('File Logic', () => {
...
+      await FileService.create(bucket, 'file.txt', USER.id, 'text', key, key2, size);
+      const newUpdate: IUpload = await UploadService.createUpdate(key, bucket, 'file.txt',  USER.id, key3, size).should.eventually.exist;
+      expect(newUpdate).to.exist;
+      expect(newUpdate.bucket).to.be.equal(bucket);
+      expect(newUpdate.name).to.be.equal('file.txt');
+      expect(newUpdate.key).to.be.equal(key);
+    });
+
+    it('should return file not exist', async () => {
```
Comment (Shahar-Y) [inline src/file/file.service.spec.ts:177]: should throw FileNotFoundError error*

File: src/file/file.service.spec.ts [hunk 4]
```diff
@@ -154,6 +154,50 @@ describe('File Logic', () => {
...
+      await UploadService.createUpdate(testUpload.key, testUpload.bucket, testUpload.name, USER.id, null, 256)
+      .should.eventually.be.rejectedWith(FileNotFoundError);
+    });
+
+    it('should throw exceeded quota error', async () => {
+      const key: string = KEY;
+      const key2: string = KEY;
+      const key3: string = KEY;
```
Comment (Shahar-Y) [inline src/file/file.service.spec.ts:185]: same

File: src/file/file.service.spec.ts [hunk 5]
```diff
@@ -154,6 +154,50 @@ describe('File Logic', () => {
...
+      await FileService.create(bucket, 'file.txt', USER.id, 'text', key, key2, size);
+      await UploadService.createUpdate(key, bucket, 'file.txt',  USER.id, key3, 101 * 1024 * 1024 * 1024)
```
Comment (Shahar-Y) [inline src/file/file.service.spec.ts]: make the `101 * 1024 * 1024 * 1024` a constant at the beginning of the file, and call it `ExceedQuotaSize`

File: src/file/file.service.spec.ts [hunk 6]
```diff
@@ -154,6 +154,50 @@ describe('File Logic', () => {
...
+        .should.eventually.be.rejectedWith(QuotaExceededError);
+    });
+
+    it('should increase quota usage by the size of the upload', async () => {
+      const oldQuota = await QuotaService.getByOwnerID(USER.id);
+      const key: string = KEY;
+      const key2: string = KEY;
+      const key3: string = KEY;
```
Comment (Shahar-Y) [inline src/file/file.service.spec.ts:195]: same

File: src/file/file.service.spec.ts [hunk 7]
```diff
@@ -154,6 +154,50 @@ describe('File Logic', () => {
...
+      expect(newUpload).to.exist;
+      expect(newUpload.bucket).to.be.equal(bucket);
+      expect(newUpload.name).to.be.equal(testUpload.name);
+      expect(newUpload.key).to.be.equal(key);
+
+      const newQuota = await QuotaService.getByOwnerID(USER.id);
+      expect(newQuota.used).to.be.equal(oldQuota.used + newUpload.size);
+    });
+  });
+
```
Comment (Shahar-Y) [inline src/file/file.service.spec.ts:207]: Add more tests: 
 - Negative size of update.
 - Update file with size other than 0.
 - Create two updates with same `key, bucket`
 - Create two updates with same `name, parent, owner`

File: src/file/file.service.ts [hunk 8]
```diff
@@ -129,7 +129,27 @@ export class FileService {
...
       await this.checkAdoption(fileId, parentID);
     }
-    return FilesRepository.updateById(fileId, partialFile);
+    const update: boolean = await FilesRepository.updateById(fileId, partialFile);
+
+    if (update) {
+      if (partialFile.size) {
+        this.updateQuota(fileId, partialFile.size);
+      }
+    }
```
Comment (Shahar-Y) [inline src/file/file.service.ts:138]: if (update && partialFile.size)...

File: src/upload/upload.model.ts [hunk 9]
```diff
@@ -36,6 +36,15 @@ export const uploadSchema: Schema = new Schema(
       type: Number,
       required: true,
       default: 0,
+    },
+    isUpdate: {
+      type: Boolean,
+      required: false,
+      default: false
+    },
+    fileID: {
```
Comment (Shahar-Y) [inline src/upload/upload.model.ts:45]: add comment: `// Relevant for updates`

File: src/upload/upload.service.ts [hunk 10]
```diff
@@ -37,20 +38,59 @@ export class UploadService {
...
+      sizeCalculated = size - file.size;
+    }
+
+    const upload: IUpload = await UploadRepository.create({ key, bucket, name, ownerID, parent, size: sizeCalculated, isUpdate: true, fileID: file.id });
+
+    if (upload) {
+      if (sizeCalculated > 0) {
+        await QuotaService.updateUsed(ownerID, sizeCalculated);
+      }
+    }
```
Comment (Shahar-Y) [inline src/upload/upload.service.ts:90]: use && to prettify

File: src/file/file.service.ts [hunk 11]
```diff
@@ -129,7 +129,27 @@ export class FileService {
...
+
+  /**
+   * Update a file quota using a subtract between new file size to the old file size.
+   * @param fileId - the id of the file.
+   * @param size - the size of the new file.
+   */
+  public static async updateQuota(fileId: string, size: number) {
+    const file: IFile = await FilesRepository.getById(fileId);
+    if (file) {
+      await QuotaService.updateUsed(file.ownerID, size - file.size);
```
[the remark this answers is not in the document]
Comment (yosef32) [reply src/file/file.service.ts:151]: זה בסדר כי אז זה מחסיר את ההפרש

File: src/file/file.service.spec.ts [hunk 12]
```diff
@@ -154,6 +161,54 @@ describe('File Logic', () => {
...
+      expect(newUpload.key).to.be.equal(KEY);
+
+      const newQuota = await QuotaService.getByOwnerID(USER.id);
+      expect(newQuota.used).to.be.equal(oldQuota.used);
+    });
+
+    it('should increase quota usage by the size of the upload', async () => {
+      const oldQuota = await QuotaService.getByOwnerID(USER.id);
+      const newFile: IFile = await FileService.create(bucket, testUpload.name, USER.id, 'text', KEY, KEY2, size);
+      const newUpload: IUpload = await UploadService.createUpdate(KEY, bucket, testUpload.name,  USER.id, KEY, 6 * 1024).should.eventually.exist;
```
Comment (Shahar-Y) [inline src/file/file.service.spec.ts:201]: use KB, MB, GB for sizes

Comment (Shahar-Y) [inline src/file/file.service.spec.ts:201]: use KB, MB, GB for sizes

File: src/file/file.service.spec.ts [hunk 13]
```diff
@@ -154,6 +161,54 @@ describe('File Logic', () => {
...
+    });
+
+    it('should throw file not found', async () => {
+      await UploadService.createUpdate(testUpload.key, testUpload.bucket, testUpload.name, USER.id, null, 256)
+      .should.eventually.be.rejectedWith(FileNotFoundError);
+    });
+
+    it('should throw exceeded quota error', async () => {
+      await FileService.create(bucket, 'file.txt', USER.id, 'text', KEY, KEY2, size);
+      await UploadService.createUpdate(KEY, bucket, 'file.txt',  USER.id, KEY, 101 * 1024 * 1024 * 1024)
```
Comment (Shahar-Y) [inline src/file/file.service.spec.ts:181]: use KB, MB, GB for sizes

File: src/upload/upload.service.ts [hunk 14]
```diff
@@ -37,20 +38,54 @@ export class UploadService {
...
+    parent: string,
+    size: number = 0,
+  ): Promise<IUpload> {
+
+    const file: IFile = await FilesRepository.getFileInFolderByName(parent, name, ownerID);
+    if (!file) {
+      throw new FileNotFoundError();
+    }
+
+    // check if the new file size is bigger from the original file, thet for save quata for the new file
```
Comment (Shahar-Y) [inline src/upload/upload.service.ts:78]: the documentation isn't clear

File: src/file/file.service.spec.ts [hunk 15]
```diff
@@ -154,6 +161,54 @@ describe('File Logic', () => {
...
+        .should.eventually.be.rejectedWith(QuotaExceededError);
+    });
+
+    it('should stay same quota usage by the negetiv size of the upload', async () => {
```
Comment (Shahar-Y) [inline src/file/file.service.spec.ts:185]: negative*

File: src/file/file.service.spec.ts [hunk 16]
```diff
@@ -154,6 +161,54 @@ describe('File Logic', () => {
...
+    it('should return a new upload', async () => {
+      await FileService.create(bucket, 'file.txt', USER.id, 'text', KEY, KEY2, size);
+      const newUpdate: IUpload = await UploadService.createUpdate(KEY, bucket, 'file.txt',  USER.id, KEY, size).should.eventually.exist;
+      expect(newUpdate).to.exist;
+      expect(newUpdate.bucket).to.be.equal(bucket);
+      expect(newUpdate.name).to.be.equal('file.txt');
+      expect(newUpdate.key).to.be.equal(KEY);
+    });
+
+    it('should throw file not found', async () => {
```
Comment (Shahar-Y) [inline src/file/file.service.spec.ts:174]:
```suggestion
    it('should throw FileNotFoundError error', async () => {
```

File: src/file/file.service.spec.ts [hunk 17]
```diff
@@ -154,6 +161,54 @@ describe('File Logic', () => {
...
+    it('should increase quota usage by the size of the upload', async () => {
+      const oldQuota = await QuotaService.getByOwnerID(USER.id);
+      const newFile: IFile = await FileService.create(bucket, testUpload.name, USER.id, 'text', KEY, KEY2, size);
+      const newUpload: IUpload = await UploadService.createUpdate(KEY, bucket, testUpload.name,  USER.id, KEY, 6 * 1024).should.eventually.exist;
+      expect(newUpload).to.exist;
+      expect(newUpload.bucket).to.be.equal(bucket);
+      expect(newUpload.name).to.be.equal(testUpload.name);
+      expect(newUpload.key).to.be.equal(KEY);
+
+      const newQuota = await QuotaService.getByOwnerID(USER.id);
```
Comment (Shahar-Y) [inline src/file/file.service.spec.ts:207]: type

File: src/file/file.service.spec.ts [hunk 18]
```diff
@@ -154,6 +161,54 @@ describe('File Logic', () => {
...
+      expect(newQuota.used).to.be.equal(oldQuota.used);
+    });
+
+    it('should increase quota usage by the size of the upload', async () => {
+      const oldQuota = await QuotaService.getByOwnerID(USER.id);
```
Comment (Shahar-Y) [inline src/file/file.service.spec.ts:199]: type

File: src/file/file.service.spec.ts [hunk 19]
```diff
@@ -154,6 +161,54 @@ describe('File Logic', () => {
...
+    it('should stay same quota usage by the negetiv size of the upload', async () => {
+      const oldQuota = await QuotaService.getByOwnerID(USER.id);
+      await FileService.create(bucket, testUpload.name, USER.id, 'text', KEY, KEY2);
+      const newUpload: IUpload = await UploadService.createUpdate(KEY, bucket, testUpload.name,  USER.id, KEY, -500).should.eventually.exist;
+      expect(newUpload).to.exist;
+      expect(newUpload.bucket).to.be.equal(bucket);
+      expect(newUpload.name).to.be.equal(testUpload.name);
+      expect(newUpload.key).to.be.equal(KEY);
+
+      const newQuota = await QuotaService.getByOwnerID(USER.id);
```
Comment (Shahar-Y) [inline src/file/file.service.spec.ts:194]: type

File: src/file/file.service.spec.ts [hunk 20]
```diff
@@ -154,6 +161,54 @@ describe('File Logic', () => {
...
+    });
+
+    it('should throw exceeded quota error', async () => {
+      await FileService.create(bucket, 'file.txt', USER.id, 'text', KEY, KEY2, size);
+      await UploadService.createUpdate(KEY, bucket, 'file.txt',  USER.id, KEY, 101 * 1024 * 1024 * 1024)
+        .should.eventually.be.rejectedWith(QuotaExceededError);
+    });
+
+    it('should stay same quota usage by the negetiv size of the upload', async () => {
+      const oldQuota = await QuotaService.getByOwnerID(USER.id);
```
Comment (Shahar-Y) [inline src/file/file.service.spec.ts:186]: type

**Comment by Shahar-Y** [comment]
Please reply to all of the comments if done, or if there is a problem. 
Also, why is there a node_modules.tar file in the PR?

File: src/file/file.service.spec.ts (hunk 7 above)
Comment (yosef32) [reply src/file/file.service.spec.ts:211]: done

File: src/upload/upload.model.ts (hunk 9 above)
Comment (yosef32) [reply src/upload/upload.model.ts:45]: done

File: src/file/file.service.spec.ts (hunk 12 above)
Comment (yosef32) [reply src/file/file.service.spec.ts:201]: done

File: src/file/file.service.spec.ts (hunk 13 above)
Comment (yosef32) [reply src/file/file.service.spec.ts:181]: done

File: src/upload/upload.service.ts (hunk 14 above)
Comment (yosef32) [reply src/upload/upload.service.ts:78]: done

File: src/file/file.service.spec.ts (hunk 15 above)
Comment (yosef32) [reply src/file/file.service.spec.ts:185]: done

File: src/file/file.service.spec.ts (hunk 18 above)
Comment (yosef32) [reply src/file/file.service.spec.ts:199]: done

File: src/file/file.service.spec.ts (hunk 19 above)
Comment (yosef32) [reply src/file/file.service.spec.ts:194]: done

File: src/file/file.service.spec.ts (hunk 20 above)
Comment (yosef32) [reply src/file/file.service.spec.ts:186]: done

Comment (Shahar-Y) [reply src/file/file.service.spec.ts:198]: not done

File: src/file/file.service.spec.ts (hunk 18 above)
Comment (Shahar-Y) [reply src/file/file.service.spec.ts:211]: not done

File: src/upload/upload.service.ts [hunk 21]
```diff
@@ -37,20 +38,54 @@ export class UploadService {
...
+    parent: string,
+    size: number = 0,
+  ): Promise<IUpload> {
+
+    const file: IFile = await FilesRepository.getFileInFolderByName(parent, name, ownerID);
+    if (!file) {
+      throw new FileNotFoundError();
+    }
+
+    // Checks whether the new file is larger than the existing file if it is greater than sending the avg to save on quota
```
Comment (Shahar-Y) [inline src/upload/upload.service.ts:78]: wut

File: src/file/file.service.spec.ts (hunk 16 above)
Comment (yosef32) [reply src/file/file.service.spec.ts]: done

**Pull request merged**